### PR TITLE
Add docs for role.spec.allow.reason.mode

### PR DESCRIPTION
**Pull request by kopiczko**
Issue https://github.com/gravitational/teleport/issues/20164

### CODE REVIEW

File: docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx [hunk 1]
```diff
@@ -294,6 +294,44 @@ user:
...
+      reason:
+        mode: 'required'
+```
+
+If a user with "node-requester" role assigned makes an Access Request for "node-access" role or any
+resource allowed by "root-node-access" they will be required to provide a reason. This is true even
+if they are assigned to another role which allows requesting those roles/resources and doesn't have
+the reason mode set to "required". Or in other words, if there are multiple roles allowing
+requesting the same roles/search_as_roles with request mode set, "require" has a higher priority
+than "optional".
```
Comment (r0mant) [inline docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx:332]: Nit: A little shorter and still carries the point across.
```suggestion
resource allowed by "root-node-access" they will be required to provide a reason. If
a user's roleset includes multiple roles governing access requests to the same roles
and resources, "require" mode takes precedence.
```

File: docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx [hunk 2]
```diff
@@ -294,6 +294,44 @@ user:
...
+|Value|Meaning|
+|---|---|
+| `optional` | The default. The user does not need to provide a reason when making a request. |
+| `required` | The user must provide a reason when making a request. |
+
+Example:
+
+```yaml
+kind: role
+version: v6
```
Comment (r0mant) [inline docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx:313]:
```suggestion
version: v7
```

File: docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx [hunk 3]
```diff
@@ -294,6 +294,44 @@ user:
...
+
+Reason mode allows enforcing users to provide reason while making an Access Request. It only works
+in allow rules and is set with `allow.request.reason.mode`.
+
+Allowed values are:
+
+|Value|Meaning|
+|---|---|
+| `optional` | The default. The user does not need to provide a reason when making a request. |
+| `required` | The user must provide a reason when making a request. |
```
Comment (r0mant) [inline docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx:307]:
```suggestion
| `required` | The user must provide a non-empty reason when making a request. |
```

**Review by r0mant — APPROVED** [review]
lgtm but hold this off until after implementation is merged

File: docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx (hunk 2 above)
Comment (kopiczko) [reply docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx:313]: Should I change it in all examples in on this page?

Comment (r0mant) [reply docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx:313]: Yeah role v7 is the latest, we should prefer using it.

File: docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx [hunk 4]
```diff
@@ -294,6 +294,44 @@ user:
 1. If one of the user's roles includes the `reason` strategy, the user must
    provide a reason when authenticating.
 
+## Reason mode
+
+Reason mode allows enforcing users to provide reason while making an Access Request. It only works
+in allow rules and is set with `allow.request.reason.mode`.
```
Comment (nklaassen) [inline docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx:300]: nit: "Reason mode" feels weird written out in english, maybe this is better
```suggestion
## Requiring request reasons

The `allow.request.reason.mode` field controls whether a reason is required when users submit access requests.
```

**Review by nklaassen — APPROVED** [review]

File: docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx (hunk 4 above)
Comment (kopiczko) [reply docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx:300]: I don't have a strong opinion here. @r0mant ?

Comment (r0mant) [reply docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx:300]: Nic's suggestion looks good to me.

**Review by smallinsky — APPROVED** [review]

File: docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx (hunk 2 above)
Comment (kopiczko) [reply docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx]: 81b0dad5074ac2eb1c1f05c077722833ed5b655a

File: docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx (hunk 4 above)
Comment (kopiczko) [reply docs/pages/admin-guides/access-controls/access-requests/access-request-configuration.mdx]: 14574945c841dd281e458f52c6c4486612bc12db

**Pull request merged**